client/hotrod-client/src/main/java/org/infinispan/client/hotrod/DataFormat.java

```diff
@@ -141,27 +208,36 @@ public byte[] valueToBytes(Object value, int estimateKeySize, int estimateValueS
    }
 
    public byte[] valueToBytes(Object value) {
-      Marshaller valueMarshaller = resolveValueMarshaller();
-      return obj2bytes(valueMarshaller, value, valueSizePredictor);
+      return client.valueToBytes(value);
    }
 
    public <T> T keyToObj(byte[] bytes, ClassAllowList allowList) {
-      Marshaller keyMarshaller = resolveKeyMarshaller();
-      return bytes2obj(keyMarshaller, bytes, isObjectStorage, allowList);
+      return client.bytesToKey(bytes, allowList);
    }
 
    public <T> T valueToObj(byte[] bytes, ClassAllowList allowList) {
-      Marshaller valueMarshaller = resolveValueMarshaller();
-      return bytes2obj(valueMarshaller, bytes, isObjectStorage, allowList);
+      return client.bytesToValue(bytes, allowList);
+   }
+
+   public MediaTypeMarshaller server() {
+      if (server == null || server.match(client)) return null;
+
+      // If is an object storage, we hash the actual object to retrieve the segment.
+      // We return null so the client does not use anything from the server.
+      if (isObjectStorage()) return null;
+
+      return marshallerRegistry.getMarshaller(server.keyType) == null ? null : server;
```

Comment:
I would suggest putting a comment here at least to say that when there is no marshaller it will cause the lookup to possibly be inefficient.